src/librustc_typeck/check/upvar.rs

```diff
@@ -116,11 +116,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
 
         let infer_kind = if let UpvarSubsts::Closure(closure_substs) = substs {
-            if self.closure_kind(closure_def_id, closure_substs).is_none() {
-                Some(closure_substs)
-            } else {
-                None
-            }
+            self.closure_kind(closure_def_id, closure_substs).is_none().to_option(closure_substs)
```

Comment:
I would have perhaps made this into a match guard instead and collapse the `None`s into one `_ => None,`. Consider it but it's up to you :)